documentation/API-Testing-Guidelines.md

```diff
@@ -1,41 +1,288 @@
+
+* Test plans are not expected to certify or measure the quality of the implementation in terms of accuracy or veracity of the response, but to validate that the interface is compliant with the API specification and that the API is fully implemented.
+
+* When the API allows options or alternative approaches, the test plan should cover those situations and define the expected behaviour for inputs or options not supported.
+  - Implementations will need to inform in advance to the API tester about which subset of options are supported. 
+
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 
+
+Test plan must validate that providing a valid input, a success response is received and that the response body complies with the JSON-schema in the API spec.
+
+For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.
+
+
+### CRUD operations <a name="crud-operations"></a>
+
+For path resources with CRUD (Create/Read/Update/Delete) operations, some scenario should validate that the new state of the resource is coherent with the response to the operation.
+
+* For Create operations:
+  - If operation success returns the created object, validate that the properties of the response object are coherent with the request object
+  - Validate that the created object can be read by the GET operation for the item id.
 
-## Best practices and recommendations <a name="recommendations"></a>
+* For Update operations:
+  - If operation success returns the updated object, validate that the properties of the response object are coherent with the request object
+  - Validate that the updated object can be read by the GET operation for the item id, with the modified properties.
+
+* For Delete operations:
+  - Validate that trying to read the successfully deleted object, with the GET operation for the deleted item id, returns the error specified in the API spec, typically 404.
+
+### Notifications <a name="notifications"></a>
+
+For operation with implicit subscriptions:
+
+* Check that when a webhook is provided, the expected events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
+* If the API allows to update a previously provided webhook:
+  - If the `notificationUrl` is modified, validate that events are received in the modified value.
+  - If the `notificationUrl` can be nullified, validate that events are not longer received.
+  - If the `notificationAuthToken` is modified, validate that the new token is used for the events.
+
+For the explicit subscriptions model:
+
+* Validate that the subscribed events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
```

Comment:
@jlurien  Consideration of new template  event-subscription-template.yaml needed in Testing-Guidelines as well

Reply:
I don't get this comment, sorry. Could you please elaborate further? Which template and which considerations?

Reply:
@jlurien https://github.com/camaraproject/Commonalities/pull/189 This was new template introduced & consideration of this for Testing guidelines will be useful cc @akoshunyadi

Reply:
@jlurien The template for explicit event subscription (like geofencing): https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml

Reply:
It should be aligned now after @PedroDiez's suggestions. Please double check